### ISSUE TITLE
Replays regularly show wrong accuracy & judgement counts

### ISSUE THREAD

**Issue by 120-cell** [ppy/osu#28744]
### Type

Game behaviour

### Bug description

The accuracy and other stats shown at the end of replays do not match the ones I got when setting the score (which are correctly displayed online). The replay acc is usually too high and stays constant upon rewatching the replay. This is not a rare occurrence, it happens with almost every replay that I look at. See my top plays below for example.

### Screenshots or videos

![image](https://github.com/ppy/osu/assets/120076941/fb4cff0f-27ca-434f-af89-34b7fbc489a3)
![image](https://github.com/ppy/osu/assets/120076941/fca31e29-d3c1-4208-a73c-70e50c99a04f) should be 97.57%, 121x 100, 3x 50
![image](https://github.com/ppy/osu/assets/120076941/c1c52017-7aa2-42ef-9651-b31d72da2f74) should be 94.82%, 68x 100, 1x 50
![image](https://github.com/ppy/osu/assets/120076941/76e54b87-d0c5-4c8a-9eed-f21587ade5b8) should be 97.08%, 26x 100, 4x 50
![image](https://github.com/ppy/osu/assets/120076941/d130fddd-4cd3-46bf-b4cf-b76a93878e96) this one is correct
![image](https://github.com/ppy/osu/assets/120076941/367e5e03-8735-4f37-b0f4-26ac22bc0ff7) should be 96.65%, 27x 100

### Version

2024.625.2

### Logs

[compressed-logs.zip](https://github.com/user-attachments/files/16103043/compressed-logs.zip)

**Comment by smoogipoo** [comment]
Sticking to just the first one, do you have the replay for that first one in your _local_ leaderboard, with the correct stats? If so can you provide that?

**Comment by 120-cell** [comment]
The screenshots are from my local replays, meaning the local replay is incorrect.  Here's the file
[120-cell playing -45 - Crimsonic dimension (Sanch-KK) [Phantasm] (2024-05-20_22-29).zip](https://github.com/user-attachments/files/16107787/120-cell.playing.-45.-.Crimsonic.dimension.Sanch-KK.Phantasm.2024-05-20_22-29.zip)
I also checked the online replay and it seems to be the same as the local one.

**Comment by ssz7-ch2** [comment]
duplicate of https://github.com/ppy/osu/issues/11311

**Comment by smoogipoo** [comment]
@ssz7-ch2 Have you actually tested to make sure that this is that issue, or are you just making a similarity inference?

**Comment by ssz7-ch2** [comment]
> @ssz7-ch2 Have you actually tested to make sure that this is that issue, or are you just making a similarity inference?

I should've added "most likely" before it as I am making a similarity inference. But I'm not sure how it would be possible to test if these specific replay discrepancies are caused by that issue. Since you won't you need the unrounded timestamps for the clicks? which I don't think is possible to retrieve anymore.

Afaik that is the only issue that causes differences in 300s/100s/50s for non-classic std plays, so I think it's very likely that that's the cause of the discrepancies seen here.

**Comment by Wieku** [comment]
I was doing lazer replay playback tests in danser and I got wrong results. I've decided to double check in lazer and noticed a discrepancy here as well:
local score:
![image](https://github.com/user-attachments/assets/820aeacb-0617-4f43-9f38-c223c9bec5b0)
1269x great, 119x ok

after replay playback:
![image](https://github.com/user-attachments/assets/3da7884d-ed13-4655-aff3-762ba8bfcf30)
1261x great, 127x ok

Replay [Wiek playing KASAI HARCORES - Cycle Hit (Worminators) [Home Run] (2024-09-22_01-16).zip](https://github.com/user-attachments/files/17155633/Wiek.playing.KASAI.HARCORES.-.Cycle.Hit.Worminators.Home.Run.2024-09-22_01-16.zip)

I've added an artificial 1ms to great hit window in danser and I got 1276x great 112x ok so it seems to be replay frame rounding.

**Comment by smoogipoo** [comment]
As a bit of an update, peppy and I briefly discussed tending to this issue as a somewhat priority-1 urgency.

The gist of it is that we could break the replay format and store frame times as `double`s. The idea being that lazer replays don't need to be compatible with stable - that's something we did while assuming stable would eventually show lazer scores in the leaderboard but that hasn't panned out in practice.

However, there are also other paths that we could go down. For example, we could keep the existing frame signature and instead [append](https://github.com/ppy/osu/blob/616c2aeefc0f086024e5b3f5130466a21eba3643/osu.Game/Scoring/Legacy/LegacyScoreEncoder.cs#L105) the `HitResult` that occurred to the file, and fake it during gameplay. This could improve more than just replay accuracy - we would also be able to have instant accurate seeks and likely simplify the server-side structures that currently serialise [internal scoring values](https://github.com/ppy/osu/blob/616c2aeefc0f086024e5b3f5130466a21eba3643/osu.Game/Rulesets/Scoring/ScoreProcessor.cs#L592-L632) to achieve the same. (In fact I wonder why we need the rest of the data in `FrameHeader` and can't just use the `Statistics`...).

We'd need to consider that it probably won't work well if we change gameplay and a result is no longer valid for the hit object type, but I'm not sure what the plans are if that changes anyway - invalidate scores?

At the end of the day, this needs some experimentation/an MVP implementation, but tl;dr we're not against breaking backwards compatibility if required.

**Comment by felipemarins** [comment]
Thank you for looking into this issue. I only have a few questions about your comment. I hope I'm not misunderstanding it, but wouldn't "instant accurate seek" break the current method of opening a replay and skipping it to the end to see the extended score statistics at the result screen? As I understand it, the replay player wouldn't process the actual hits between the current frame and the final one, so there wouldn't be a way to get the stats. I get that this isn't an optimal way either, but I guess it's something to consider.

Also, as much as I think it would be nice if lazer replays could work on stable, wouldn't it be incompatible with lazer-only mods or mod settings (such as rate adjust) anyway?

**Comment by smoogipoo** [comment]
I consider that process a hack that should be a background operation. Related: the PP community has asked about storing more info in scores that could be used in PP calc, and that could be fulfilled by replays...

**Comment by tsunyoku** [comment]
I think I'd want to be careful with involving PP in this. I think the general direction of including the timed statistics in replays is good, I'm not commenting on that.

However, I have concerns that if we start encouraging changes to PP that use those timed statistics we may result in an extremely unbalanced system as we'd have to consider and separately handle scores that don't have access to this data (for example, there are many legacy scores with no replays so there is absolutely *zero* chance that we could have this data for those) and I don't imagine this ever working well in practice - it'd be unbalanced, and also a very confusing experience.

This is before even thinking about any possible concerns with the individual ideas that people may have that would involve these extra statistics... I'm not saying that the committee have a hard "no" stance to using these in PP but at a glance I don't see it working so I'm wary of encouraging it. This is just my personal opinion, I'm not sure how the other committee members feel about it and it's probably also not very relevant at this stage - I just see PP mentioned and feel the need to leave my opinion.

**Comment by stanriders** [comment]
I'd rather have more timed info in the scores themselves so that it can be clearly separated into the "has new data" and "has not" categories that would be more obvious for everyone. Using replays would muddy everything very hard very quickly to the point that people might have different pp on identical scores - for example if one player set it when the map got ranked, and other set it  5 years later. 
I think the only way replays could be used in PP if _all_  replays were preserved, not necessarily from the beginning of the game. I'm definitely in favor of extending score data by any means though, even without populating all legacy scores. My only concern is consistency

**Comment by ssz7-ch2** [comment]
I think storing frame times as doubles should be done first. Then other methods of storing replay data can be considered later on.

Since currently that info is permanently lost, so replays that are inaccurate today will be inaccurate forever.

**Comment by Givikap120** [comment]
> I think storing frame times as doubles should be done first. Then other methods of storing replay data can be considered later on.
> 
> Since currently that info is permanently lost, so replays that are inaccurate today will be inaccurate forever.

It's already stored as doubles
```c#
/// <summary>
/// The time at which this <see cref="ReplayFrame"/> takes place.
/// </summary>
[Key(0)]
public double Time;
```

The question is that is it stored as double on the server. Because exporting the replay (using `LegacyEncoder`) will definitely truncate inputs to int.

**Comment by pigeonskew** [comment]
When will this get fixed? It's really off-putting when you watch the replay, and you'd get high accuracy on the replay than the recorded score. Like... what should I believe in? Makes me feel like I deserve more score/pp than the recorded score since in the replay, I performed better. 

This is another one of those cases where it's hard to main lazer. The reliability just isn't there yet.

**Comment by peppy** [comment]
agree that this is pretty game breaking high priority. will bump up and we can assess where to slot this in to the ongoing development.

**Comment by serobor** [comment]
https://osu.ppy.sh/beatmapsets/2257930#osu/4804490

I've got SS rank, but after watching replay, it shows 1x100

![Image](https://github.com/user-attachments/assets/76aad21a-0825-412e-8c42-22c4f78f4404)
![Image](https://github.com/user-attachments/assets/7b4cdcfc-9ef8-4ea9-a015-9dfbbefe0046)

**Comment by chrispycone** [comment]
Also experiencing this bug. Except in my case, the game won't give me the correct judgement on just one note in the map.
https://osu.ppy.sh/scores/4919915126

This is me playing the map:

https://github.com/user-attachments/assets/5d3123b4-d8e3-46b6-a1a3-34f31c4bf0ea

This is how it shows up in replay:

https://github.com/user-attachments/assets/c6835a3d-c8b3-4305-9b2e-9e3c444c6a1f

Why does this judgement gets treated like its too early? 

![Image](https://github.com/user-attachments/assets/3585cb5a-7c68-42ca-a761-997234c94833)

**Comment by Theighlin** [comment]
Hate to be that person, but is there anything left to address this issue?
Old replay obviously can't be fixed, but I was under the impression that after #33148, #33491 (and maybe others i've missed) new ones are supposed to be accurate, is that correct?

I've played on 2025.702.0 and can still produce plays that play back a tiny bit inaccurately, and can provide them with a liveplay comparison if needed.

Sorry if there are still things to be done to address this that i'm not aware of.

**Comment by bdach** [comment]
@Theighlin please provide evidence, including the resulting replays

**Comment by Theighlin** [comment]
Play 1

https://osu.ppy.sh/scores/5094109753

left is live, right is replay
https://youtu.be/XblZmw9vmj0
(one discrepancy is seen at 0:29)

[Theighlin playing Mili - Rightfully (Okoratu) \[Screw the Dice\] (2025-07-03_10-04).zip](https://github.com/user-attachments/files/21038278/Theighlin.playing.Mili.-.Rightfully.Okoratu.Screw.the.Dice.2025-07-03_10-04.zip)


Play 2
https://osu.ppy.sh/scores/5094860690

![Image](https://github.com/user-attachments/assets/0eb8381d-bfed-40ec-b082-f28ec4b83a75)

i don't know how to say this, but i can't make a comparison for this one because seemingly the original recording does not proceed at a consistent speed, so it's impossible to line up with the replay (but it's here is the original play https://youtu.be/F0TfWnSr704)

[Theighlin playing Hoshimachi Suisei - KING (-Aqua) \[Queen\] (2025-07-03_14-08).zip](https://github.com/user-attachments/files/21038189/Theighlin.playing.Hoshimachi.Suisei.-.KING.-Aqua.Queen.2025-07-03_14-08.zip)

I noticed when jumping to the end of a replay that scrubbing the replay at different speeds yields different results, moving frame-by-frame fixes the discrepancy seen in the video (pay attention to the animation of the sliderend with the number 6.)
Also why is the sliderend miss "x" appearing if the judjment gets increased and the follow circle closes in instead of expanding?

https://github.com/user-attachments/assets/59571b9b-3ced-4185-97e9-187e1fefa9ff

**Comment by Ladwick** [comment]
Same for me, the amount of times this happen has decreased since the fix for sure but it still keeps happening

Getting 96.62% on replay and 96.82% on results
[Video proof](https://www.youtube.com/watch?v=BwuveNwklt8)
[Online score](https://osu.ppy.sh/scores/5091850031)
Also, I'm on 2025.702.0-tachyon
![Image](https://github.com/user-attachments/assets/abe963e3-d3c0-4a3b-8ff3-0215305eae12)

**Comment by bdach** [comment]
"thankfully", for my sanity at least, the extent of the remaining inaccuracies pointed out here seems contained to slider end treatment, which is one area that I haven't really touched much (except for the important frame change I guess), and judging by the fact that the replays sometimes play correctly and sometimes not, may be not an issue that permanently corrupts replays. I've had to play some of these multiple times to get the discrepancy.

somewhat tempted to close this issue and spin whatever that last reported issue is off to a separate one but I dunno. I'll think about it after I investigate more.

**Comment by bdach** [comment]
I've moved the latest issue into its own separate thread at https://github.com/ppy/osu/issues/34016.

I'm going to make a call to close & lock this thread because I feel if that it stays open it'll become an uncontrollable mess of everyone flowing in and posting that they have the "same issue" that isn't _really_ the same issue anymore going forward. Treat this as a clean break after the batch of recent replay-related fixes which should have eliminated a bulk of these occurrences.

Anyone else from the future coming here to report a new replay inaccuracy issue - please open a separate issue, it will be triaged, diagnosed, and handled appropriately by me.
